### PR TITLE
Fix bug that log can't be shown when task runs failed

### PR DESCRIPTION
**Pull request by huozhanfeng**
---
The log can't be shown normally when the task runs failed. Users can only get useless logs as follows. #13692

<pre>
*** Log file does not exist: /home/airflow/airflow/logs/dag_id/task_id/2021-06-28T00:00:00+08:00/28.log
*** Fetching from: http://:8793/log/dag_id/task_id/2021-06-28T00:00:00+08:00/28.log
*** Failed to fetch log file from worker. Unsupported URL protocol 
</pre>

The root cause is that scheduler will overwrite the hostname info into the task_instance table in DB by using blank str in the progress of `_execute_task_callbacks` when tasks into failed.  Webserver can't get the right host of the task from task_instance because the hostname info of  task_instance table is lost in the progress.

### CODE REVIEW

**Comment by uranusjr** [comment]
Can a test be added for this?

**Comment by huozhanfeng** [comment]
> Can a test be added for this?

Sure, I will try to append tests for it.

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -1495,8 +1495,7 @@ def handle_failure(
             except Exception:
                 self.log.exception('Failed to send email to: %s', task.email)
 
-        if not test_mode:
-            session.merge(self)
+        session.merge(self)
```
Comment (uranusjr) [inline airflow/models/taskinstance.py]: Is this related?

File: tests/dag_processing/test_processor.py [hunk 2]
```diff
@@ -643,6 +643,30 @@ def test_execute_on_failure_callbacks(self, mock_ti_handle_failure):
...
+        with create_session() as session:
+            session.query(TaskInstance).delete()
+            dag = dagbag.get_dag('example_branch_operator')
+            task = dag.get_task(task_id='run_this_first')
+
+            ti = TaskInstance(task, DEFAULT_DATE, State.RUNNING)
+            ti.hostname = "test_hostname"
+
+            session.add(ti)
+            session.commit()
```
Comment (uranusjr) [inline tests/dag_processing/test_processor.py]: Split this block into two `create_session()`s instead of using `commit()`.

**Review by uranusjr — COMMENTED** [review]
Logic LGTM! A couple of minor comments/questions.

I wonder if/how we can refactor `SimpleTaskInstance` to prevent this kind of error in the future. Right now, if a field is added to `TaskInstace`, `SimpleTaskInstace` and `_execute_task_callbacks` needed to be checked to pass them on. This can be difficult to maintain.

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (huozhanfeng) [reply airflow/models/taskinstance.py]: Yes, the original logic can't write task_instance info into DB/session in ut mode.

**Comment by huozhanfeng** [comment]
> Logic LGRM! A couple of minor comments/questions.
> 
> I wonder if/how we can refactor `SimpleTaskInstance` to prevent this kind of error in the future. Right now, if a field is added to `TaskInstace`, `SimpleTaskInstace` and `_execute_task_callbacks` needed to be checked to pass them on. This can be difficult to maintain.

Agree with you, maybe we can raise an independent pr to solve it.

File: tests/dag_processing/test_processor.py (hunk 2 above)
Comment (huozhanfeng) [reply tests/dag_processing/test_processor.py]: Sure

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (uranusjr) [reply airflow/models/taskinstance.py]: But I assume there is a reason it doesn’t save the task instance, although I do not know what. What would fail if you remove this change?

**Comment by uranusjr** [comment]
Please see comments above: https://github.com/apache/airflow/pull/16768#discussion_r663619814

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (huozhanfeng) [reply airflow/models/taskinstance.py:1498]: > But I assume there is a reason it doesn’t save the task instance, although I do not know what. What would fail if you remove this change?

I think so at the beginning, and I have tried to find the historical git log, nothing found. The UT always successful if remove this change although the logic is wrong.

Comment (huozhanfeng) [reply airflow/models/taskinstance.py]: @uranusjr Find the one.  Looks like it's related to the airflow test command. In the latest UT, UT will raise `IndexError: list index out of range` with the code block `assert tis[0].hostname == "test_hostname"` if remove this change.  

I wonder whether it is reasonable to mix UT and test commands together here by using a test_mode variable because the test command doesn't want to save the state into DB/session but UT should save the state into DB/session. Do you have any idea about it?
<pre>
commit aa01fab4906df3175e5e6589ec0095a5586279db
Author: Maxime Beauchemin <maxime.beauchemin@apache.org>
Date:   Mon Jan 12 21:54:40 2015 +0000

    Improving the test command line subcommand

https://github.com/apache/airflow/commit/aa01fab4906df3175e5e6589ec0095a5586279db
</pre>

**Comment by huozhanfeng** [comment]
> Please see comments above: [#16768 (comment)](https://github.com/apache/airflow/pull/16768#discussion_r663619814)

Sure, but I meet a problem with UT.

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (huozhanfeng) [reply airflow/models/taskinstance.py]: I add a config `dag_file_processor.UNIT_TEST_MODE = False` in UT as a workaround. It works well but I wonder whether it is normative.

**Comment by huozhanfeng** [comment]
> Please see comments above: [#16768 (comment)](https://github.com/apache/airflow/pull/16768#discussion_r663619814)

Resolved it by using a tricky method...

**Comment by uranusjr** [comment]
The solution is good to me actually, not tricky at all but what it’s supposed to be done 🙂 Let’s see how the test suite feels about this.

**Comment by huozhanfeng** [comment]
> The solution is good to me actually, not tricky at all but what it’s supposed to be done 🙂 Let’s see how the test suite feels about this.

That's great and thank you!

**Review by uranusjr — APPROVED** [review]

**Comment by huozhanfeng** [comment]
Hello, Should I rebase the PR to continue the CI? 

`Please rebase or amend, and force push the PR to run full tests`

File: airflow/dag_processing/processor.py [hunk 3]
```diff
@@ -573,6 +573,7 @@ def _execute_task_callbacks(self, dagbag: DagBag, request: TaskCallbackRequest):
                 ti.start_date = simple_ti.start_date
                 ti.end_date = simple_ti.end_date
                 ti.try_number = simple_ti.try_number
+                ti.hostname = simple_ti.hostname
```
Comment (kaxil) [inline airflow/dag_processing/processor.py]: probably calling `ti.refresh_from_db()` can do the trick instead of this?

Have you tested with this fix?

Comment (huozhanfeng) [reply airflow/dag_processing/processor.py]: Good idea, I will have a try.

Comment (huozhanfeng) [reply airflow/dag_processing/processor.py]: Resolved it and let the `ti.refresh_from_db()` action only runs in the failure_callback block due to performance concern.

Comment (huozhanfeng) [reply airflow/dag_processing/processor.py]: Actually, I think  #16804 is the better way to solve this problem. Sometimes DB resource is tight.

File: airflow/dag_processing/processor.py [hunk 4]
```diff
@@ -568,14 +568,11 @@ def _execute_task_callbacks(self, dagbag: DagBag, request: TaskCallbackRequest):
...
-                ti.start_date = simple_ti.start_date
-                ti.end_date = simple_ti.end_date
-                ti.try_number = simple_ti.try_number
-                ti.state = simple_ti.state
-                ti.test_mode = self.UNIT_TEST_MODE
                 if request.is_failure_callback:
+                    ti = TI(task, simple_ti.execution_date)
+                    #TODO use simple_ti to improve performance here in the future
+                    ti.refresh_from_db()
+                    ti.test_mode = self.UNIT_TEST_MODE
```
Comment (uranusjr) [inline airflow/dag_processing/processor.py]: `handle_failure_with_callback` has an explicit `test_mode` argument, you can use that instead of hacking the value on like this

Comment (huozhanfeng) [reply airflow/dag_processing/processor.py]: Sure, I will improve it now.

File: airflow/dag_processing/processor.py (hunk 3 above)
Comment (huozhanfeng) [reply airflow/dag_processing/processor.py]: @kaxil hello, could you please help to take a look?

File: tests/dag_processing/test_processor.py [hunk 5]
```diff
@@ -643,6 +643,30 @@ def test_execute_on_failure_callbacks(self, mock_ti_handle_failure):
                 test_mode=conf.getboolean('core', 'unit_test_mode'),
             )
 
+    def test_failure_callbacks_should_not_drop_hostname(self):
+        dagbag = DagBag(dag_folder="/dev/null", include_examples=True, read_dags_from_db=False)
+        dag_file_processor = DagFileProcessor(dag_ids=[], log=mock.MagicMock())
+        dag_file_processor.UNIT_TEST_MODE = False
+
+        with create_session() as session:
+            session.query(TaskInstance).delete()
```
Comment (ephraimbuddy) [inline tests/dag_processing/test_processor.py]:
```suggestion
```
We have `clear_db_runs` in the setup method which clears TaskInstance as well, so we don't need this

Comment (huozhanfeng) [reply tests/dag_processing/test_processor.py]: Sure, thank you.

**Review by ephraimbuddy — APPROVED** [review]

**Review by ephraimbuddy — APPROVED** [review]

**Comment by huozhanfeng** [comment]
hello, the failed checks seem to have nothing to do with this feature, I wonder how to merge it into the repo?

**Review by kaxil — APPROVED** [review]
